src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -174,40 +161,43 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                     {
                         // If there's no razor code in this app, don't do anything.
                         RazorSourceGeneratorEventSource.Log.DiscoverTagHelpersFromReferencesStop();
-                        return ImmutableArray<TagHelperDescriptor>.Empty;
+                        return null;
+                    }
+
+                    // If there aren't any references, we can bail out early.
+                    // Note: compilation.References just enumerates compilation.ExternalReferences and compilation.DirectiveReferences.
+                    if (compilation.ExternalReferences is [] && compilation.DirectiveReferences is [])
```

Comment:
It avoids the enumerator allocation when calling `compilation.References`.

Reply:
Right, but when would we actually observe this optimization in practice? I would expect the only time this would ever be true is for `System.Private.Corlib`, and they don't use this source generator so they'd not get any benefits.

Reply:
Could it be observed in scenarios where the project system is still adding files/references to a compilation? I'm not certain when source generators run.

Reply:
I don't think so, though I'll let @chsienki double check me on that.

Reply:
At any rate, I'm fine removing this, since you seem to feel more strongly than I do.

Reply:
Thanks for humoring me 🙂.